File: it.unibo.bls19Local/src/main/java/it/unibo/bls/appl/MainBlsMockBase.java
package it.unibo.bls.appl;

import it.unibo.bls.devices.ButtonType;
import it.unibo.bls.devices.DeviceFactory;
import it.unibo.bls.devices.LedType;
import it.unibo.bls.interfaces.*;
import it.unibo.bls.applLogic.BlsApplicationLogic;			//JAVA IMPL
//import it.unibo.bls.kotlin.applLogic.BlsApplicationLogic;	//KOTLIN IMPL
import it.unibo.bls.listener.ButtonObserver;
import it.unibo.bls.utils.Utils;
import it.unibo.bls.interfaces.IButtonListener;


public class MainBlsMockBase  {
	protected IButtonObservable btn;
	protected ILed led;
	protected IButtonListener buttonObserver;
	protected IControlLed applLogic;
	protected DeviceFactory devFactory = new DeviceFactory();
	protected LedType ledType ;
	protected ButtonType buttonType;

//Factory method
  	public static MainBlsMockBase createTheSystem(){
  		return new MainBlsMockBase();
 	}
//Not-visible Constructor
 	protected MainBlsMockBase( ) {
		Utils.showSystemInfo();
		setDeviceTypes();
		createTheComponents();
		connectTheComponents();
		startTheSystem();
 	}
 	protected void setDeviceTypes(){
		ledType    = LedType.LedMockObj;
		buttonType = ButtonType.ButtonMockObj;
	}
 	protected void createTheComponents(){
  		led       = devFactory.createLed( ledType );
		applLogic = new BlsApplicationLogic();
 		btn       = devFactory.createButton( buttonType );
 		led.turnOff();
		buttonObserver = ButtonObserver.createButtonListener( );
	}
	protected void connectTheComponents(){
		buttonObserver.setControl( applLogic );
		applLogic.setControlled(led);
	}

	protected void startTheSystem(){
  		btn.addObserver( buttonObserver );	//starts the job
	}

	//Selectors (introduced for testing)
 	public IButtonObservable getButton(){
		return btn;
	}
	public ILed getLed(){
		return led;
	}
	public IButtonListener getButtonObserver(){
		return buttonObserver;
	}
	public IControlLed getApplLogic(){
  		return applLogic;
	}
/*
MAIN
*/
	public static void main(String[] args) {
   		MainBlsMockBase sys = createTheSystem();
 	}
}

File: it.unibo.bls19Local/src/main/kotlin/it/unibo/bls/kotlin/appl/MainBlsMockBase.kt
package it.unibo.bls.kotlin.appl

import it.unibo.bls.kotlin.applLogic.BlsApplicationLogic
import it.unibo.bls.kotlin.devices.ButtonType
import it.unibo.bls.kotlin.devices.DeviceFactory
import it.unibo.bls.kotlin.devices.LedType
import it.unibo.bls.kotlin.interfaces.IApplListener
import it.unibo.bls.kotlin.interfaces.IButtonObservable
import it.unibo.bls.kotlin.interfaces.IControl
import it.unibo.bls.kotlin.interfaces.ILed
import it.unibo.bls.kotlin.listener.ButtonObserver
import it.unibo.bls.kotlin.utils.Utils


class MainBlsMockBase//Not-visible Constructor
private constructor() {
    //Selectors (introduced for testing)
    var button: IButtonObservable? = null  //refactored
    var led: ILed? = null
    var actionListener: IApplListener? = null
    var applLogic: IControl? = null
        //get(){return field}
        //set(value){ field = value }
    val cmd = "click"
    private val devFactory = DeviceFactory()

    init {
        createTheComponents()
        connectTheComponents()
    }

    protected fun createTheComponents() {
        led = devFactory.createLed(LedType.LedMockObj)
        applLogic = BlsApplicationLogic()
        button = devFactory.createButton(ButtonType.ButtonMockObj)
        led!!.turnOff()
        actionListener = ButtonObserver.createButtonListener()
    }

    protected fun connectTheComponents() {
        println("KOTLIN VERSION connectTheComponents  ")
        actionListener!!.setControl(applLogic!!)
        (applLogic as BlsApplicationLogic).setTheLed(led!!)  //type casting
        button!!.addObserver(actionListener!!)    //starts the job
    }

    companion object {
        //Factory method
        fun createTheSystem(): MainBlsMockBase {
            return MainBlsMockBase()
        }

/*
MAIN
*/
        @JvmStatic
        fun main(args: Array<String>) {
            println("KOTLIN VERSION")
            createTheSystem()
            Utils.delay(5000)
        }
    }
}